std/stdio.d

```diff
@@ -1542,6 +1549,173 @@ for every line.
+
+        static string genTempName(const(char)[] prefix, const(char)[] suffix, const(char)[] dir)
+        {
+            import std.ascii : digits, letters;
+            import std.random : rndGen;
+            import std.range : chain;
+            import std.string : representation;
+            auto name = (char[]).init ~ prefix ~ "012345678901234" ~ suffix;
+
+            // Replace the 012345678901234 with random characters
+            rndGen.popFront();
+            auto chars = chain(letters.representation, digits.representation);
+            foreach (ref c; name[prefix.length .. $ - suffix.length])
+            {
+                c = chars[rndGen.front % chars.length];
+                rndGen.popFront();
+            }
+
+            return buildPath(dir, name);
+        }
+
+        // Once should be enough given the randomness, but we'll give it three
+        // times to be safe.
+        foreach (i; 0 .. 3)
+        {
+            auto filename = genTempName(prefix, suffix, dir);
+
+            // So that we can test the case where the file already exists.
+            version (unittest)
+            {
+                if (i == 0)
+                {
+                    import std.file;
+                    std.file.write(filename, "nothing");
+                }
+            }
+
+            // O_EXCL does not support symbolic links.
+            if (dirName(filename).isSymlink)
+                filename = buildPath(absolutePath(dirName(filename), baseName(filename)));
+
+            import std.internal.cstring : tempCString;
+            version (Posix)
+            {
+                import core.sys.posix.sys.stat, core.sys.posix.fcntl;
+                auto fd = open(tempCString(filename), O_RDWR | O_CREAT | O_EXCL, S_IRUSR | S_IWUSR);
+            }
+            else version (Windows)
+            {
+                version (DIGITAL_MARS_STDIO)
+                {
+                    enum openFlags = O_RDWR | O_CREAT | O_EXCL;
+                    enum permissions = S_IREAD | S_IWRITE;
+                }
+                else version (MICROSOFT_STDIO)
+                {
+                    enum openFlags = _O_RDWR | _O_CREAT | _O_EXCL;
+                    enum permissions = _S_IREAD | _S_IWRITE;
+                }
+                auto fd = _wsopen(tempCString!wchar(filename), openFlags, _SH_DENYNO, permissions);
+            }
+            else static assert("Unsupported OS");
+
+            if (fd == -1)
+            {
+                version (unittest)
+                {
+                    import std.file;
+                    if (i == 0 && filename.exists)
+                        std.file.remove(filename);
+                }
+                continue;
+            }
+
+            File retval;
+            retval.fdopen(fd, "w+b", filename);
+            return retval;
+        }
+
+        throw new Exception("Failed to create a temporary file.");
```

Comment:
Okay. It now throws an `ErrnoException` for the last failed open attempt. So, it won't tell you why the others failed, but you could check the error code of the last one to see whether it was a permissions problem or whatever. I'm not sure what else you might want me to do here. The only other thing that I can think of would be to create a new exception type, which would be overkill IMHO.